Create "2021 Deview Chapter.details" post (#48)
makinarocks/makinarocks.github.io: _posts/2021-11-17-floorplan-automation-part-II.md

---
layout: post
title: Floorplan Automation - Part II
author: wooshik myung
categories: [combinatorial_optimization, reinforcement_learning]
image: assets/images/2021-11-17-Deview-details/technical_overview.png
---

<!-- 
Outlines
1. 문제에 대한 간단한 설명
2. 문제의 의의 (@jiyoonlim123 포스팅 레퍼런싱)
3. 문제를 풀기위해 선택한 방법(접근): 현업자들의 어떤 요구사항을 바탕으로 이러이러한 방법론을 채택함
4. 강화학습에 대한 아주아주 간단한 설명: State로부터 Action을 결정하고, 환경은 이에 대한 Reward를 반환. Reward의 기댓값을 최대화 시키는 형태로 목적함수 정의 -> 그렇다면 우리는 State / Action / Reward를 어떻게 정의했나?
5. State (Clustering, Feature 정보)
6. Action (Grid Cell이라는 이산공간 정의, 셀의 배치순서)
7. Reward에 대한 설명 (HPWL / Congestion)
8. 그렇다면 State로부터 Action을 어떻게 도출하나? (Model - GNN + CNN): Preprocessing을 여기서 언급
9. 실험 세팅 (강화학습 Agent / 장비 / ...)
10. 실험 결과
11. Further Works 
-->

<script>
    var figure_idx = 1
    function func()
    {
        document.write(figure_idx)
        figure_idx = figure_idx + 1
    }
</script>

# Related to

> * [Deview 영상](https://tv.naver.com/v/23650123)

Reviewer: Deview21 발표영상: "발표제목..."

* [포스팅 1부](@jiyoon)

# 프로젝트 소개

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/placements.gif" alt="placement animation" width="100%">
</p>

> </figure>

Reviewer: 그림에 대한 Description이 추가되면 더 좋겠습니다.

> 본 프로젝트는 지난 2020년 6월 구글에서 발표해 Nature 등재로 화제가 되었던 **A graph placement methodology for fast chip design**[[1](#ref-1)[2](#ref-2)] 논문에 기초를 두고 있습니다.

Reviewer: ```suggestion
본 프로젝트는 지난 2020년 6월 구글에서 발표해 Nature 등재로 화제가 되었던 **A graph placement methodology for fast chip design**[[1](#ref-1)][[2](#ref-2)] 논문에 기초를 두고 있습니다.
```

> 저희 프로젝트 역시 Application-Specific Integrated Circuit(ASIC) 반도체의 제조 공정 중 하나인 Floorplan 단계에서의 Optimizaition 문제를 강화학습으로 해결하고자 합니다.

Reviewer: ASIC 반도체 or ASIC ?

Floorplan은 소자들을 배치하는 과정 중 하나이며, 트랜지스터가 아닌 블록단위의 배치를 말합니다.
수십~수백만개 단위의 트랜지스터와의 연결성을 고려하며 블록들을 배치해야 하며, 작업의 특성상 현업 전문가가 하더라도 시간이 많이 소요되는 작업 중하나입니다.
문제 정의부터 반도체 산업의 지식을 많이 녹여야 하는 만큼, 이해하고 구현까지의 과정에서 많은 어려움이 있었습니다.

> 프로젝트를 진행하며 겪은 약 5개월간의 시간동안 우리가 이 문제를 어떻게 정의하고 해결해가고 있나를 설명드리겠습니다.

Reviewer: ```suggestion
프로젝트를 진행하며 겪은 약 5개월간의 시간동안 저희가 이 문제를 어떻게 정의하고 해결해가고 있나를 설명드리겠습니다.
```

먼저 아래에 나오는 용어들은 다음과 같은 뜻을 가지고 있습니다.
* Macro: Floorplan과정에서 배치되는 블록, 메모리 구성 요소로 Standard Cell 대비 훨씬 큰 크기를 가지고 있음.
* Standard Cell: NAND, NOR, XOR과 같은 논리 게이트, 크기가 매우 작으며 수십~수백만개 단위의 소자가 사용됨.
* Net: Macro와 Standard Cell들의 연결관계를 정의하는 Wire.
* Netlist: Net들의 집합체.

# Is that Real Problem?
Floorplan 과정은 일반적으로 사람의 손에 의해 이루어집니다.

> Chip Canvas라는 2차원의 연속 공간에서 복잡한 연결관계를 고려하며 수백개에 달하는 Macro들을 배치하는 작업은 큰 탐색공간(Search Space)로 인해 높은 복잡도를 가지고 있는 문제입니다.

Reviewer: ```suggestion
Chip Canvas라는 2차원의 연속 공간에서 복잡한 연결관계를 고려하며 수백개에 달하는 Macro들을 배치하는 작업은 큰 탐색공간(Search Space)으로 인해 높은 복잡도를 가지고 있는 문제입니다.
```

> 또한, 직접 Macro를 배치한 후 그 뒤의 설계과정은 상용 Electronic Design Automation(EDA) Tool들을 통해 수행되고 나서야 배치에 대한 평가를 내릴 수 있습니다.

Reviewer: ```suggestion
또한, 직접 Macro를 배치한 후 그 뒤의 설계과정은 상용 Electronic Design Automation(EDA) Tool을 통해 수행되고 나서야 배치에 대한 평가를 내릴 수 있습니다.
```

> 만약 결과가 좋지 않다면, Macro를 배치하는 작업부터 다시 수행해야 합니다.
> 이런 방식이다 보니, 숙련된 기술자라도 며칠에서 몇주의 시간이 소요되기도 합니다.

Reviewer: ```suggestion
만약 평가 결과가 좋지 않다면, Macro를 배치하는 작업부터 다시 수행해야 하기 때문에 숙련된 기술자라도 며칠에서 몇 주의 시간을 소요하기도 합니다.
```

기술자가 반복적인 경험을 통해 습득한 지식을 기반으로 진행되기 때문에, 숙련도에 따라서 필요한 시간이 크게 달라지는 작업 중 하나입니다.
현업에서 종사하시는 전문가분들과의 협업을 통해 Floorplan의 주요 문제점은 아래와 같이 정리해볼 수 있겠습니다.

* Macro의 배치를 수작업으로 진행 후, EDA Tool을 이용해 결과를 확인하는 작업을 여러번 반복해야하는 만큼 시간이 오래걸림.
* 사람의 경험이 매우 중요하기 때문에, 숙련도에 따라 품질과 요구되는 시간의 편차가 큼.
  
연구 배경 및 의의에 대한 더욱 자세한 설명은 관련 포스팅을 참고해주세요.(@jiyoon.lim)
이 포스팅에서는 문제정의와 해결방법 관련된 기술적인 내용을 중적적으로 설명합니다.

# How to solve?

> 기존의 Placement Optimization Problem에 적용되는 많은 Classical한 방법들(Simulated Annealing[[2](#ref-3)], Genetic Algorithm[[3](#ref-4)]...)은 좋은 성능을 보여주고 있지만, 새로운 문제를 만나면 같은 도메인이더라도 초기 상태부터 다시 최적화를 진행해야한다는 단점을 가지고 있습니다.

Reviewer: 앞선 포스팅에서는 Placement를 STD Cell을 배치하는 과정으로 소개

한번 학습된 모델이 다르게 정의된 문제에서 별다른 추가 학습 없이 어느 수준의 이상의 성능을 가지며 문제를 푸는 것을 일반화(Generalize)라고 부릅니다.
예를 들어, 소자 5개, 10개짜리 Netlist를 이용해 학습하면 20개 규모의 Netlist에 대해서도 일정 수준 이상의 성능이 나오는 것입니다.
이미 많은 연구들에서 신경망을 직접시킨 강화학습은 강력한 Generalize 수준을 보여주고 있습니다.
강화학습의 이러한 강점을 이용한다면 **여러 구조와 크기의 Netlist에 대해서 추가적인 학습 없이 또는 최소한의 학습시간만**을 가지고 좋은 배치를 찾을 수 있게됩니다.
저희는 이점에 주목하여 강화학습을 주요 알고리즘으로써 사용했습니다.
사람이 경험을 통해 습득하여 반복적으로, 직관적으로 수행하는 업무를 데이터 기반으로 학습하여 해결하는 것의 최종목표입니다.
아래에서는 먼저 강화학습이 무엇인지에 대해 간단히 설명드리고 저희가 이 문제에 대해 어떻게 접근하고 있는지 설명합니다.

### 강화학습이란?

> 강화학습은 에이전트(Agent)가 환경(Environment)과 상호작용하며 시행착오(Trial and Error)를 통해 더 좋은 보상(Reward)를 얻기위해 좋은 행동(Action)을 학습하는 알고리즘입니다.

Reviewer: ```suggestion
강화학습은 에이전트(Agent)가 환경(Environment)과 상호작용에서의 시행착오(Trial and Error)를 통해 더 높은 보상(Reward)를 얻을 수 있는 좋은 행동(Action)을 학습하는 알고리즘입니다.
```

## Action Decision
Graph Feature Embedding 된 정보를 하나의 이미지로써 인식해 다음 소자가 배치될 확률분포를 계산합니다. 
마지막 레이어의 계산 결과에 Invalid Action Masking*이 적용되며, 해당 값들을 이용해 소자의 위치를 결정합니다.
이 과정은 아래 그림처럼 표현됩니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/action_decision.png" alt="policy network" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Policy Network</figcaption>
</p>
</figure>

Feature Embedding을 CNN에 입력하여 점진적으로 채널을 1까지 줄입니다. 
Input Channel이 점진적으로 줄어드는 과정에서 Receptive Field는 커지므로, 결국에는 전 영역을 모두 볼 수 있게끔 조절할 수 있습니다. 
마지막 Layer의 출력은 현재 소자를 배치할 위치에 대한 확률분포를 나타냅니다. 
에이전트는 해당 확률분포에 따라 행동을 결정합니다.

# Experiment Setup & Details
### Dataset
저희는 실험을 위해 두가지 Netlist를 사용하였습니다. 상세 스펙은 아래와 같습니다(by Auto Placement of ICC2).

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/dataset.png" alt="dataset specification" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Dataset specification</figcaption>
</p>

> </figure>

Reviewer: incorrect orientation of RAMWRAPPER

> Toy Netlist는 중간중간 떠오르는 아이디어나 모델들이 잘 동작하는지 확인하기 위해 사용했습니다. 

Reviewer: RAMWRAPPER 포스팅에 포함할지 여부 확인

실질적인 성능 검증 및 아래에 기술할 내용은 약 12만개의 소자들로 구성된 오른쪽 Netlist를 사용하여 얻은 결과입니다. 

> GNN 모델은 Residual Connection을 사용해 다양한 크기의 그래프에 대해 동일한 모델을 사용하더라도 준수한 성능을 낼 수 있는 **DeeperGCN**[[6](#ref-7)]을 사용했습니다. 

Reviewer: DeeperGCN 사용에 대한 당위성 설명이 구구절절할 것 같음 -> GCN 모델로 표현하면 어떨지?

<!-- ### Experiment Setup
아래 설명드릴 실험 모두 A100 한대를 사용한 기준으로 설명합니다. 
사용한 오픈소스 및 프로그램은 아래와 같습니다.   -->

<!-- * Ray - 분산학습을 위해 사용했습니다.
* Rllib - 다양한 Agent를 적용해보기 위해 사용되었으며, 설명드릴 실험 결과에서는 PPO를 사용했습니다.
* torch_geometric - GNN모델을 쉽게 적용하기 위해 사용했습니다.
* kahypar - hmetis 구현을 위해 사용했습니다.
* ICC2 - RL에서 학습한 결과를 human expert, auto placement와 비교해 보기 위해 사용했습니다.   -->

<!--
## Invalid episode handling
해당 실험에서는 Toy Netlist를 이용해 실험하였습니다. 
가장 먼저 Invalid episode 발생률을 알아보기 위해 여러 grid size에서 랜덤으로 1000번의 배치를 진행해보았습니다. 
얻은 실험 결과 중, 아래의 baseline을 사용해 실험을 진행합니다.

|   grid    |   Valid Episodes  |
|   :--:    |   :------------:  |
|   6X6     |   32.4%           |

### Constant Penalty
<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/constant_penalty.png" alt="normal gradient" width="100%">
  <figcaption style="text-align: center;">[그림1] - Dataset specification</figcaption>
</p>
</figure>
### Overlap penalty  
### Big Canvas  
-->

# Evaluations
실험의 난이도를 줄일 수 있는 두가지 휴리스틱(Heuristic)을 적용했습니다.

#### Single Cluster
강화학습의 주요 목적은 복잡한 연결관계를 고려하여 최적의 Macro 배치를 찾는 것입니다.

> Standard Cell에 대한 의존성을 줄이고 최대한 간단하게 대략적인 위치를 예측하기 위해서 해당 휴리스틱을 적용했습니다.

Reviewer: 대략적인 STD Cell 위치에 대한 예측을 휴리스틱으로 사용

모든 Standard Cell들을 하나의 Cluster로 묶어 배치를 수행하기 때문에 Episode 당 Step 수가 줄어들어 학습을 쉽게 만듭니다.

#### Bounding Box
일반적으로 Standard Cell끼리는 높은 연결 밀도를 가지고 있습니다.
그 때문에, 실제 칩배치 전문가들이 배치한 결과들을 살펴보면 대부분의 소자배치문제에서 Macro는 Chip Canvas의 바깥쪽 영역에, Standard Cell은 중앙쪽에 위치하고 있습니다.

> 이러한 방식에 착안하여, Macro와 Standard Cell에 대해 개별적인 Invalid Masking 방식을 적용합니다.

Reviewer: ```suggestion
이러한 방식에 착안하여, Macro와 Standard Cell에 대해 개별적인 배치영역을 사용합니다.
```

학습초기에 Bouding Box를 설정하면 해당 영역에 Macro를 배치하지 않습니다.

> 해당 방법 역시 탐색공간이 줄어들어 학습을 쉽게 만드는 효과를 얻을 수 있습니다.

Reviewer: ```suggestion
그럼으로써 탐색공간을 줄여 학습을 쉽게 만드는 효과를 얻을 수 있습니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/vanilla_rl.png" alt="placment and training curve" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - 별도의 휴리스틱을 적용하지 않은 배치결과와 학습 곡선 </figcaption>
</p>
</figure>

'소자들을 가까이 배치'만하면 되는 HPWL은 직관적인 만큼 초반부터 빠르게 감소하는 모습을 보여줍니다.
반대로, 초반에 Congestion값이 오르는 것으로 보아 Routing Path가 특정 지역에 몰리는 것은 잘 파악하지 못합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/training_flow.png" alt="placment at 10 epochs and 800 epochs" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - 10 Epochs의 배치 결과와 800 Epochs에서의 배치 결과 </figcaption>
</p>
</figure>

학습 초반부에는 무작위로 배치되던 소자들이 중반부로 갈수록 Standard Cell은 중앙으로, Macro는 바깥에 위치하는 모습을 보여줍니다.
이는 전문가의 배치나, 툴을 이용한 Auto Placement와도 상당히 흡사한 모습을 보여줍니다.
학습 중후반에는 Standard Cell들끼리 위치를 바꿔가며 Congestion을 최적화합니다.
학습 그래프를 보면, 직관적인 HPWL을 먼저 줄이고 세밀한 위치 교환을 진행하며 Congestion을 줄이는 것을 알 수 있습니다.
이 경우에는 약 1000 Epochs 후에 수렴합니다.

## 이 결과가 얼마나 좋은거지?
사실 위의 결과들은 단순히 에이전트의 학습 과정만을 보여줄 뿐, 실제로 어느정도의 수준인지는 보여주지 않습니다.

> 위의 결과가 실제로 좋은건지 확인하기 위해서는 ICC2와 같은 EDA Tool을 이용해 측정해야합니다.

Reviewer: 독자들이 상상하기 어려울 것 같음

    <td bgcolor="Silver"> Group 3 </td>
    <td style="color:red"> <strong> 0.79 </strong> </td> <td style="color:red"> <strong> 161 </strong> </td>
    <td style="color:red"> <strong> 0.79 </strong> </td> <td style="color:red"> <strong> 161 </strong> </td>
    <td> 0.65 </td> <td> 158 </td>
    <td> 0.69 </td> <td> 158 </td>
    <td style="color:red"> <strong> 0.79 </strong> </td> <td> 160 </td>
  </tr>
  <tr align="center">
    <td bgcolor="Silver"> Group 4 </td>
    <td> 2.60 </td> <td> 227 </td>
    <td> 2.65 </td> <td style="color:red"> <strong> 230 </strong> </td>
    <td style="color:red"> <strong> 2.66 </strong> </td> <td style="color:red"> <strong> 230 </strong> </td>
    <td> 2.60 </td> <td> 227 </td>
    <td> 2.55 </td> <td style="color:red"> <strong> 230 </strong> </td>
  </tr>
  <tr align="center">
    <td bgcolor="Silver"> Group 5 </td>
    <td> 2.67 </td> <td> 231 </td>
    <td style="color:red"> <strong> 2.74 </strong> </td> <td style="color:red"> <strong> 235 </strong> </td> 
    <td> 2.73 </td> <td> 234 </td>
    <td> 2.68 </td> <td> 231 </td>
    <td> 2.63 </td> <td> 234 </td>
  </tr>
</table>

Floorplan단계에서 가장 주의깊게 봐야 할 수치는 WNS, FREQ 등이 있습니다.

> 일반적으로, WNS는 양수인지를 확인하고, FREQ는 높을수록 좋습니다.

Reviewer: WNS는 음수일때 Violation이므로

각 그룹은 성능 검사를 위해 미리 정해둔 Routing Path의 집합입니다.
결과적으로 **'각 그룹에 대해 0이상의 WNS와 FREQ가 높을 경우 성능이 좋다'**라고 말할 수 있겠습니다. (각 지표들에 대한 자세한 설명은 @jiyoon.lim 포스팅을 참고해주세요.)
결과를 보시면, 저희의 에이전트는 두가지 Baseline에 근접한 수치를 보여줍니다.

> 아직 개발 초기상태인걸 감안하면 유의미한 성과로 보입니다.

Reviewer: ```suggestion

```

# 앞으로 풀어야 할 숙제는?

> MakinaRocks COP팀에서 개발한 강화학습 기반의 알고리즘은 소규모 배치 문제에서 전문가 및 상용 EDA툴의 자동 배치에 근접하는 성능을 보였습니다.

Reviewer: "EDA툴의 자동 배치에 근접하는 성능" <- 아직 성능을 어필할 레벨이 아니므로 작은 규모의 문제에서 유의미한 학습결과를 보인 것으로 서술 (큰 문제를 풀기위한 준비가 되었음)

하지만 앞으로 해결해야 할 문제들이 남아있습니다.

* 더 적은 학습시간과 더 좋은 배치 결과를 얻을 수 있는 방법은?
* 강화학습으로 배치되는 Standard Cell의 배치와 ICC2에서 배치되는 결과의 간극을 어떻게 메꿀 수 있을까?
* Generalize를 잘하려면 어떻게 해야할까?
* 학습이 완료되고 소자들을 세밀하게 움직이기 위한 Post-processing을 어떻게 구현해야 할까?

> * 현업자분들이 진정으로 원하는 기능에는 또 무엇이 있을까?

Reviewer: 질문을 줄이고 내용을 보강하면 어떨까요?
e.g. 왜 Generalization을 잘해야 하는지

반대로, Bin을 너무 크게 설정하면 충분히 세밀한 탐색을 못하게 됩니다.
또한, Bin의 크기에 따라서 소자들을 랜덤하게 배치했을 때 공간상의 제약때문에 모든 소자들을 배치하지 못하는 경우가 빈번하게 발생합니다.
그래서 저희는 특정 범위의 Bin 크기에 대해 각각 $n$번의 배치를 랜덤으로 진행해 모든 소자를 배치할 확률이 높은 Bin의 크기로 설정했습니다.

<a name="a.2."></a>

### A.2. Invalid Action Masking
현재 배치에 따라 다음 배치될 소자가 배치될 수 있는 영역은 다릅니다. 
가장 먼저, 소자 종류에 관계없이 Chip Canvas를 벗어나면 안됩니다. 
Macro의 경우에는 다른 소자와 겹침을 허용하면 안되며, Standard Cell의 경우 배치될 면적과 밀도를 고려해 배치해야 합니다. 
이러한 특성들을 고려해 Action을 효과적으로 제어하기 위한 작업이 Invalid Action Masking입니다. 
아래 그림은 이 예시를 보여줍니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/invalid_action_mask.png" alt="Invalid Action Masking" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Invalid Action Masking</figcaption>
</p>
</figure>

## B. Reward Function Design
<a name="b.1."></a>

### B.1. Invalid Episode Handling
소자들을 배치하다 보면, 모든 소자들이 배치되기 전 모든 Bin들에 배치 불가능한 상태에 놓이는 경우가 있습니다. 
Macro는 Overlapping을 아예 허용하지 않다 보니, 큰 Macro 소자가 들어갈 공간이 부족해 생기는 현상입니다. 
가장 먼저 소자들을 크기순으로 배치하면 이러한 현상이 발생하는 확률을 줄일 수 있습니다. 
하지만, 배치 순서를 정하는 것만으로는 완전히 해결할 수 없으며, 에이전트에게 배치가 완료되지 않는 상태가 최악의 결과라는 정보를 전달해야 합니다.

> Invalid Episode가 발생할 경우, $-\alpha-\beta$ 값의 Penalty Reward를 반환합니다.

Reviewer: Wirelength와 Congestion을 [0,1]로 바운드했기 때문에 -alpha - beta가 Reward의 하한이 됨

에이전트는 보상의 총합을 높이기 위해 Invalid Episode가 발생하지 않는 방향으로 학습합니다.

<!-- 이 문제를 해결하기 위해 여러가지 아이디어들이 제시되었으며 내부적으로 실험을 진행한 결과, Invalid Episode가 발생하면 $-\alpha-\beta$ 값의 Penalty를 주는 방식으로 해결하였습니다. -->
<!-- (\*해당 아이디어에 대한 자세한 내용들은 현재 특허 출원 진행 중이며, 출원이 완료되면 추후 업데이트하겠습니다.) -->

<!--
하지만, 이 방법은 invalid episode를 효과적으로 handling하기 어려우므로, 이를 해결하기 위한 방법들은 아래와 같이 나열해 볼 수 있겠습니다.

1. Overlapping을 허용 후, Reward function 설계시, Overlap된 면적에 대한 Penalty 부여
2. 큰 Canvas Size에서 시작해, 기존의 Canvas Size에서 벗어나는 소자 개수 혹은 면적에 대해 Penalty 부여
3. Invalid Episode가 발생할 경우, 큰 Penalty(constant) 를 부여하고 해당 episode 종료
   
3가지 방법 모두 실험해 본 결과, 3번째 방법이 가장 좋은 성능을 보여주었습니다. 
결과에 따라, 아래에서 소개드릴 실험부분의 내용들은 3번을 이용한 결과입니다.
-->

<a name="b.2."></a>

### B.2. Normalization & Reward Coefficient
위에서 설정한 WireLength와 Congestion을 적절하게 조절하기 위해서는 Normalization과 Coefficient를 활용해야 합니다. 
먼저 두개의 값을 일정 Boundary안에 넣어주기 위해 Normalization을 진행합나디.
#### WireLength
HPWL의 최대값은 Net의 두 소자가 Chip Canvas의 대각선 양 끝단에 위치하는 경우입니다. 
각 Net에 대해 HPWL을 계산한뒤 최대값으로 나누어주면 효과적으로 [0, 1] 사이의 값으로 바운딩이 가능합니다.
#### Congestion
Congestion의 경우 WireLength보다는 약간 더 복잡합니다. 

> MST혹은 Steiner Tree를 통해 생성된 Edge들이 어떤 Bin을 지날 때, 그 유형은 아래와 같이 4가지가 있습니다.

Reviewer: ```suggestion
Minimum Spanning Tree 혹은 Steiner Tree를 통해 생성된 Edge들이 어떤 Bin을 지날 때, 그 유형은 아래와 같이 4가지가 있습니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/congestion_normalize.png" alt="4 paths of edge" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Bin을 지나가는 Edge의 4가지 유형</figcaption>
</p>
</figure>

4가지 유형 모두 Horizontal, Vertical Congestion을 구해 평균값을 취해보면 0.5가 나오는 것을 알 수 있습니다. 
즉, 최악의 경우 한 Net의 모든 Edge들이 한개의 Bin을 지나는 것이라 생각해 볼 수 있을 것입니다. 
이 경우의 Congestion 값은 `0.5 X n_edges`가 되므로 각 Net에 대한 Congestion을 해당 값으로 나누면 역시 [0, 1] 사이의 값으로 바운딩이 가능합니다. 

> WireLength와 Congestion에 대해 각각의 Coefficient를 곱해주는 방식으로 둘 사이의 Scaling을 진행합니다. 

Reviewer: 빼거나 아니면 Reward Formulation을 하단에 추가하거나

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/congestion_filter.png" alt="Convolution Filter" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Convolution Filter 적용 전(좌)과 후(우)</figcaption>
</p>
</figure>

위의 그림을 보시면 Standard Cell Cluster에 과도하게 집중된 Congestion값(그림에서 S라 표시된 Bin)이 효과적으로 완화된 것을 보실 수 있습니다. 
이 작업에 대한 추가적인 이점으로는, Congestion을 주변 영역으로 확장함으로 인해서 Clustering을 진행함으로써 발생한 원본 데이터와의 오차도 줄일 수 있다는 장점도 가지고 있습니다. 

# References
<a name="ref-1">[1]</a>  [Azalia Mirhoseini, Anna Goldie, Mustafa Yazgan, Joe Jiang, Ebrahim Songhori, Shen Wang, Young-Joon Lee, Eric Johnson, Omkar Pathak, Sungmin Bae, Azade Nazi, Jiwoo Pak, Andy Tong, Kavya Srinivasa, William Hang, Emre Tuncer, Anand Babu, Quoc V. Le, James Laudon, Richard Ho, Roger Carpenter, Jeff Dean, 2021, A graph placement methodology for fast chip design.](https://www.nature.com/articles/s41586-021-03544-w)

<a name="ref-2">[2]</a>  [Google AI, 2020, Chip Placement with Deep Reinforcement Learning, Google AI Blog.](https://ai.googleblog.com/2020/04/chip-design-with-deep-reinforcement.html)

> <a name="ref-3">[2]</a>  [S. Kirkpatrick, C. D. Gelatt, Jr., M. P. Vecchi, 1983, Optimization by Simulated Annealing.](https://www.science.org/doi/abs/10.1126/science.220.4598.671)

Reviewer: should be 3

---
layout: post
title: Floorplan Automation - Part II
author: wooshik myung
categories: [combinatorial_optimization, reinforcement_learning]
image: assets/images/2021-11-17-Deview-details/technical_overview.png
---

<!-- 
Outlines
1. 문제에 대한 간단한 설명
2. 문제의 의의 (@jiyoonlim123 포스팅 레퍼런싱)
3. 문제를 풀기위해 선택한 방법(접근): 현업자들의 어떤 요구사항을 바탕으로 이러이러한 방법론을 채택함
4. 강화학습에 대한 아주아주 간단한 설명: State로부터 Action을 결정하고, 환경은 이에 대한 Reward를 반환. Reward의 기댓값을 최대화 시키는 형태로 목적함수 정의 -> 그렇다면 우리는 State / Action / Reward를 어떻게 정의했나?
5. State (Clustering, Feature 정보)
6. Action (Grid Cell이라는 이산공간 정의, 셀의 배치순서)
7. Reward에 대한 설명 (HPWL / Congestion)
8. 그렇다면 State로부터 Action을 어떻게 도출하나? (Model - GNN + CNN): Preprocessing을 여기서 언급
9. 실험 세팅 (강화학습 Agent / 장비 / ...)
10. 실험 결과
11. Further Works 
-->

<script>
    var figure_idx = 1
    function func()
    {
        document.write(figure_idx)
        figure_idx = figure_idx + 1
    }
</script>

# Related to

> * [Naver Deview 21 발표영상 \[TITLE: 강화학습, 산업의 난제에 도전하다! - ASIC 반도체 설계 (Floorplan) 자동화\]](https://tv.naver.com/v/23650123) 

Reviewer: 제목이 먼저 나오면 어떨까 싶습니다.
```suggestion
* [강화학습, 산업의 난제에 도전하다! - ASIC 반도체 설계 (Floorplan) 자동화 (Naver Deview 21 발표영상)](https://tv.naver.com/v/23650123) 
```

> * [포스팅 1부](@jiyoon)

Reviewer: 추후 포스팅 제목이 정해지면 제목으로 수정 부탁드립니다.

# 프로젝트 소개

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/placements.gif" alt="placement animation" width="100%">
  <figcaption style="text-align: center;">
  <p style="font-weight: bold;"> FIGURE ABSTRACT </p>
  위 애니메이션은 약 1500 iterations 정도의 학습 과정을 보여주고 있습니다.
  녹색의 큰 사각형은 Macro, 청색의 작은 사각형은 Standard Cell Cluster를 나타냅니다. 
  어느 정도 학습이 진행된 후, 사각형의 움직임이 눈에 띄게 줄어들며 안정된 모습을 보여줍니다.
  에이전트는 Macro는 Chip Canvas 영역의 바깥쪽, Standard Cell Cluster는 안쪽에 위치시키도록 학습합니다.
  이에 대한 자세한 설명은 본 포스팅을 참고해주세요.
  </figcaption>
</p>
</figure>

----
<p></p>

#### Introduction

본 프로젝트는 지난 2020년 6월 구글에서 발표해 Nature 등재로 화제가 되었던 **A graph placement methodology for fast chip design**[[1](#ref-1)][[2](#ref-2)] 논문에 기초를 두고 있습니다.
강화학습을 이용해 주문형 반도체 (Application-Specific Integrated Circuit, ASIC)의 제조 공정 중 하나인 Floorplan 단계에서의 Optimization 문제를 해결하고자 합니다.
Floorplan은 소자들을 배치하는 과정 중 하나이며, 트랜지스터가 아닌 블록 단위의 배치를 말합니다.

> 수십~수백만 개 단위의 트랜지스터와의 연결성을 고려하며 블록들을 배치해야 하기 때문에 시간이 많이 소요되는 작업 중 하나입니다.
> 문제 정의부터 반도체 산업의 지식을 많이 녹여야 하는 만큼, 이해하고 구현까지의 과정에서 많은 어려움이 있었습니다.
> 프로젝트를 진행하며 겪은 약 5개월간의 시간 동안 저희가 이 문제를 어떻게 정의하고 해결해가고 있나를 설명하겠습니다.

Reviewer: Simple is the best
```suggestion
수십~수백만 개 단위의 트랜지스터와의 연결성을 고려하며 블록들을 배치해야 하기 때문에 시간이 많이 소요되는 작업 중 하나입니다.
본 포스팅에서는 약 5개월간 경험한 Floorplan 문제의 이해부터 솔루션을 구현까지의 과정을 공유드리겠습니다.
```

먼저 본 포스팅에서 자주 언급되는 주요 용어들과 뜻입니다.

> * Macro: Floorplan 과정에서 배치되는 블록, 메모리 구성 요소로 Standard Cell 대비 훨씬 큰 크기를 가지고 있음.

Reviewer 2: 파트 1에서의 설명과 같이 Macro 도 블록보다 명확하게 소자라는 단어를 사용하면 더 좋을 것 같습니다.

* Standard Cell: NAND, NOR, XOR과 같은 논리 게이트, 크기가 매우 작으며 수십~수백만 개 단위의 소자가 사용됨.
* Net: Macro와 Standard Cell들의 연결 관계를 정의하는 Wire.
* Netlist: Net들의 집합체.

# Is that Real Problem?
Floorplan 과정은 일반적으로 사람의 손에 의해 수행됩니다.
Chip Canvas라는 2차원의 연속 공간에서 복잡한 연결 관계를 고려하며 수백 개에 달하는 Macro들을 배치하는 작업은 큰 탐색공간(Search Space)으로 인해 높은 복잡도를 가지고 있는 문제입니다.
또한, 직접 Macro를 배치한 후 그 뒤의 설계과정을 상용 Electronic Design Automation(EDA) Tool을 통해 수행하고 나서야 배치에 대한 평가를 할 수 있습니다.
만약 평가 결과가 좋지 않다면, Macro를 배치하는 작업부터 다시 수행해야 하므로 숙련된 기술자라도 며칠에서 몇 주의 시간을 소요하기도 합니다.
반복적인 경험을 통해 습득한 지식을 기반으로 진행되는 Floorplan은 기술자의 숙련도에 따라서 필요한 시간이 크게 달라지곤 합니다.
현업에서 종사하시는 전문가분들과의 협업을 통해 알게 된 Floorplan의 주요 문제점은 아래와 같이 정리할 수 있습니다.

* Macro의 배치를 수작업으로 진행 후, EDA Tool을 이용해 결과를 확인하는 작업을 여러 번 반복해야 하는 만큼 시간이 오래 걸림.
* 사람의 경험이 매우 중요하기 때문에, 숙련도에 따라 품질과 요구되는 시간의 편차가 큼.

연구 배경 및 의의에 대한 더욱 자세한 설명은 관련 포스팅을 참고해주세요. (@jiyoon.lim)
이 포스팅에서는 문제정의와 해결 방법 관련된 기술적인 내용을 중점적으로 설명합니다.

# How to solve?
기존의 배치 최적화 문제*에 적용되는 많은 Classic한 방법들(Simulated Annealing[[3](#ref-3)], Genetic Algorithm[[4](#ref-4)]...)은 좋은 성능을 보여주고 있지만, 새로운 문제를 만나면 같은 도메인이라도 초기 상태부터 다시 최적화를 진행해야 한다는 단점이 있습니다.

> 한번 학습된 모델이 다르게 정의된 문제에서 별다른 추가 학습 없이 어느 수준 이상의 해답을 찾는 것을 일반화(Generalize)라고 합니다.
> 예를 들어, 소자 5개, 10개짜리 Netlist를 이용해 학습하면 20개 규모의 Netlist에 대해서도 일정 수준 이상의 결과를 얻을 수 있습니다.

Reviewer: Minor comment
```suggestion
한번 학습된 모델이 다르게 정의된 문제에서 별다른 추가 학습 없이 어느 수준 이상의 해답을 찾는 것을 일반화(Generalize)라고 합니다.
소자 5개, 10개로 구성된 Netlist를 학습하여 20개 규모의 Netlist에 대해서도 일정 수준 이상의 결과를 얻어내는 것이 그 예시입니다.
```

이미 많은 연구에서 신경망을 직접 시킨 강화학습은 강력한 Generalize 수준을 보여주고 있습니다.
강화학습의 이러한 강점을 이용한다면 **여러 구조와 크기의 Netlist에 대해서 추가적인 학습 없이 또는 최소한의 학습 시간만**을 가지고 좋은 배치를 찾을 수 있습니다.
저희는 이점에 주목하여 강화학습을 주요 알고리즘으로써 사용했습니다.

> 사람이 반복적인 경험을 통해 직관적으로 수행하는 업무를 데이터 기반으로 학습하여 해결하는 것의 최종목표입니다.

Reviewer: 서론에 의해 독자들이 여러가지 Netlist에 대한 일반화 실험 결과를 기대하게 될 것 같습니다.
저희는 단일 Netlist만을 실험했기 때문에 이 점을 충분히 서론에서 주지시킬 필요가 있어 보입니다.
(e.g. 단일 Netlist 학습을 시작으로 차후에는 일반화 학습으로 나아갈 것을 암시)

아래에서는 먼저 강화학습이 무엇인지에 대해 간단히 기술하고 저희가 이 문제에 대해 어떻게 접근하고 있는지 설명합니다.

\*  <font size="-1"> 배치 최적화 문제: 대상의 물리적인 위치를 통해 목표함수를 최적화하는 문제 </font>

### 강화학습이란?
강화학습은 에이전트(Agent)가 환경(Environment)과 상호작용에서의 시행착오(Trial and Error)를 통해 더 높은 보상(Reward)을 얻을 수 있는 행동(Action)들을 학습하는 알고리즘입니다.
환경은 에이전트의 행동마다 다음 상태(State)와 보상을 반환하며, 에이전트는 전달받은 상태를 고려해 다시 행동을 결정하게 됩니다.

> 강화학습은 **Markov Decision Process**[[5](#ref-5)] (MDP)에 기반해, 차례로 어떤 의사결정(decision process)을 해야 하는 문제들에 많이 적용됩니다.

Reviewer: ```suggestion
강화학습은 **Markov Decision Process**[[5](#ref-5)] (MDP)에 기반해, 차례로 어떤 의사결정(Decision Process)을 해야 하는 문제들에 많이 적용됩니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/mdp.png" alt="markov decision process" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Markov Decision Process</figcaption>
</p>
</figure>

결국 에이전트는 학습하는 과정 중에 Sampling 되는 모든 에피소드(Episode)*에 대한 보상의 기댓값을 최대화하기 위한 행동들을 학습합니다.

> 당연하게도, 강화학습의 구성 요소(에이전트, 상태, 행동, 보상)들의 정의방법에 따라 다른 학습양상을 보여줍니다.

Reviewer: 이 부분에서는 주어를 명확하게 기술하는 것이 앞 문장과의 연결을 고려할때 좀 더 문장이 매끄러워 보입니다.

```suggestion
당연하게도, 강화학습은 그 구성 요소들(에이전트, 상태, 행동, 보상)의 정의방법에 따라 다른 학습양상을 보여줍니다.
```

아래에서는 저희가 환경과 에이전트를 어떻게 설계했는지 설명드리겠습니다.

\*  <font size="-1"> Episode: 에이전트가 일련의 행동들을 수행하고 보상들을 얻는 Sequence, 에피소드가 종료되면 환경은 초기상태로 리셋된다. </font>

### Technical Overview
각 단계의 자세한 내용을 설명해 드리기 앞서 환경을 어떻게 정의하고 모델을 어떻게 설계했는지에 대해 전체적인 흐름을 설명합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/technical_overview.png" alt="technical overview" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Technical Overview</figcaption>
</p>

> </figure>

Reviewer: 다이어그램에서 Agent를 확대하여 모델의 구조를 보여주고 있는데,
환경으로부터 Macro Features와 Netlist Graph를 입력받는 것으로 독자가 오해할 수 있을 것 같습니다.

Reviewer: 모델 그림에서 몇몇 명칭 아리에 붉은 선이 보이네요.

##### 환경 정의
에이전트는 환경의 주요 구성요소인 상태, 행동, 보상을 통해 환경과 상호작용하며 학습합니다.
그러므로 에이전트를 효과적으로 학습시키기 위해서는 각 요소를 적절히 정의해야 합니다.
저희의 경우 각 요소는 아래와 같은 역할을 수행해야 합니다.

* 상태: 행동을 결정하기 위한 판단에 필요한 정보를 담고 있어야 함.

> * 행동: 더 좋은 Chip Design을 더 적은 복잡도로 얻을 수 있게 하는 행동의 집합.

Reviewer: ```suggestion
* 행동: 더 좋은 Chip Design을 더 낮은 복잡도로 얻을 수 있게 하는 행동의 집합.
```

Reviewer: `더 낮은 복잡도`라는 것이 어떤 의미인가요?
이 문제에서 에이전트의 행동에 따라 복잡도가 변경되는 부분이 있나요?

Author: 해당 부분 삭제했습니다!

* 보상: Chip Design의 품질.

저희가 설계한 환경은 아래와 같은 순서로 진행됩니다.

> 1. 빈 Chip Canvas인 초기 상태에서 시작하여 행동 한 번에 소자 한 개를 배치

Reviewer 2: 마지막에 마침표를 추가해야 할 것 같습니다.

> 2. 배치 중에는 0을 반환

Reviewer: 마지막 배치도 '배치 중'으로 볼 수 있으므로 좀 더 명확히 하면 어떨까 합니다.
e.g. 마지막 배치 외에는 모든 배치에 대해 0의 보상을 반환

이를 구현하기 위해, Graph Neural Network(GNN)과 Convolutional Neural Network(CNN)으로 이루어진 모델을 설계했습니다.
각각의 목적은 아래와 같습니다.

* GNN: Graph 형태의 Netlist의 정보를 효과적으로 Embedding하기 위해 선택했습니다.
* CNN: 배치가 완료된 소자들의 지역적인 정보를 효과적으로 학습할 수 있는 CNN 구조를 선택했습니다.

이에 대한 자세한 설명 역시 아래에서 설명합니다.

# 강화학습 환경 정의하기
## State Representation
상태는 현재 환경이 어떤 상태에 있는지에 대한 정보를 담고 있습니다.
그러므로, 상태는 에이전트가 좋은 행동을 결정하기 위해 도움이 되는 정보들을 전달해야 합니다.
칩 설계의 전문가분들이 **소자들의 연결 관계**와 **물리적 특성**들을 고려해 배치하는 것에 착안하여 에이전트가 다음 행동을 결정하기 위해 필요한 정보를 아래와 같이 정의했습니다.

1. 소자들 간의 연결 관계
2. 각 소자의 물리적 정보 (width, height, is_macro, n_pins ... etc.)
3. 이번에 배치할 소자의 정보
4. 지금까지의 배치 정보

각 소자들 간의 연결 관계가 아래와 같은 Hypergraph*의 형태를 띠고 있는 만큼, 소자들의 물리적 특성들을 Node Features로 정의할 수 있습니다. 

\*  <font size="-1"> Hypergraph: Edge한개에 다수개의 Vertex가 연결되어 있는 그래프 구조 </font>

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/hypergraph.png" alt="hypergraph" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Hypergraph 예시 </figcaption>
</p>
</figure>

Graph Clustring Alogrithm을 이용하여 수십~수백만 개 단위의 Standard Cell들에 Clustering을 진행합니다.
결과적으로, 배치하려는 대상의 개수가 줄어들어 문제의 복잡도가 감소합니다.

> Clustering 과정에서의 정보손실을 최소화하기 위해 Hypergraph Clustering Algorithm인 hmetis[[6](#ref-6)] 알고리즘을 사용합니다. 

Reviewer: ```suggestion
Clustering 과정에서의 정보손실을 최소화하기 위해 Hypergraph Clustering Algorithm인 hMETIS[[6](#ref-6)] 알고리즘을 사용합니다. 
```

추후 배치 과정에서 한 Cluster에 속해있는 소자들을 묶어 하나의 독자적인 소자로 취급합니다.

## Action Representation
행동은 에이전트가 상태의 정보들(소자들의 물리적 특성과 현재 배치상태 등)을 이용해 현재 배치할 소자를 어느 위치에 둘 것인지 결정짓는 역할을 합니다. 
에이전트가 얻는 보상을 최대화하기 위해서는 충분한 탐험(Exploration, 경험)이 필요합니다.
경험을 효율적으로 하기 위해서는 적절한 크기의 탐색공간을 정의해야 합니다.
저희는 이러한 점들을 고려해 연속공간의 Chip Canvas를 2차원 Grid의 이산 공간으로 정의했습니다.

> 에이전트가 Neural Network로부터 얻어진 확률분포에서 특정 Bin을 선택하면 배치될 소자의 중앙이 해당 Bin의 중앙에 위치하도록 구현했습니다.

Reviewer: 박지우 박사님이 Grid Cell이라는 명칭이 더 자연스럽다고 하셨으니 명칭을 Grid Cell로 하면 어떨까요?
```suggestion
에이전트가 Neural Network로부터 얻어진 확률분포에서 특정 Grid Cell을 선택하면 배치될 소자의 중앙이 해당 Grid Cell의 중앙에 위치하도록 구현했습니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/action.png" alt="action" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Action </figcaption>
</p>
</figure>

저희가 구현한 방식은 탐색공간을 줄이면서 배치 결과를 평가하기 위한 Approximated Metric을 단순하게 정의할 수 있다는 장점이 있습니다.
해당 단계에서 고려되어야 할 아래 issue들에 대해 아래 [APPENDIX](#appendix)에 정리했습니다.

* [A.1. Bin Size Decision](#a.1.)
* [A.2. Invalid Action Masking](#a.2.)

## Reward Representation
보상은 에이전트에게 해당 행동이 얼마나 유익한지를 알려주는 학습에 굉장히 중요한 요소입니다. 
일반적으로 ASIC의 Floorplan 결과는 ICC2와 같은 상용 EDA Tool을 사용하여 측정됩니다. 
Floorplan 이후의 배치 결과를 평가하기 위한 추가 설계과정은 모두 해당 Tool에서 수행되기 때문에, 보통 수 시간이 소요됩니다.

> 강화학습은 보상을 반복적으로 Sampling하여 학습하는 구조이기 때문에 Tool의 결과를 직접적으로 보상으로써 정의하기에는 다소 무리가 있습니다. 

Reviewer: 보상을 sampling한다는 표현은 오해의 여지가 있어보입니다.
```suggestion
강화학습은 다량의 샘플을 통해 학습하는 방법이기 때문에 Tool의 결과를 직접적으로 보상으로써 정의하기에는 계산 속도상 다소 무리가 있습니다. 
```

> 동일 시간 내에 더 많은 Sample을 확보하기 위하여 Floorplan의 결과를 빠른 시간 내에 근사하는 함수를 설계해야 합니다.

Reviewer: ```suggestion
그렇기 때문에 동일 시간 내에 더 많은 Sample을 확보하기 위하여 Floorplan의 결과를 빠른 시간 내에 근사하는 함수를 설계해야 합니다.
```

> 하지만 그전에 먼저 좋은 배치가 무엇인지를 알아야 합니다. 

Reviewer: 내용이 전환된다기 보다는 확장되는 부분이라 `하지만`보다는 다른 표현이 적절해보입니다.
```suggestion
배치결과의 품질을 적절히 근사하려면 먼저 좋은 배치가 무엇인지를 알아야 합니다. 
```

산업에서는 일반적으로 PPA (Performance, Power, Area)라는 기준들을 사용합니다. 
각 항목에 대응되는 효과는 아래와 같습니다.

* Performance - 높은 클럭 주파수(Hz)
* Power - 적은 전력 소모
* Area - 작은 Chip Canvas

> 우리가 설계한 근사함수를 에이전트의 보상함수로 사용하기 위해서는 위의 3가지 사항이 모두 고려되어야 합니다. 

Reviewer: ```suggestion
에이전트의 보상함수로 근사함수를 사용하기 위해서는 위의 3가지 사항이 모두 고려되어야 합니다. 
```

> 소자들이 가깝게 배치될수록 데이터가 전달되는 Path의 길이가 짧아지고 자연스레 3가지 모두 해결될 것입니다.

Reviewer: 좀 더 자세하게 설명하면 어떨까요?
```suggestion
언뜻보면 소자들을 가깝게 배치하면 데이터를 전달하는 Path의 길이가 짧아지면서 더 높은 Power, Performance와 더 낮은 Area를 얻을 수 있으리라 생각할 수 있습니다.
```

이 말인즉슨, 소자끼리 Data를 주고받을 Routing Path가 필요한데 이 자원이 한정적이라는 것입니다. 
그래서 소자들을 무작정 가깝게 배치해 단위 구역에 제공되는 Routing Resources보다 많은 Routing Path가 생기면 아래 그림처럼 다른 쪽으로 우회해서 신호를 전달해야 합니다. 

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/detour.png" alt="routing detour" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Routing Detour</figcaption>
</p>
</figure>

위의 그림에서 Macro 2와 Standard Cell 5의 가장 짧은 Routing Path가 적색 선임을 쉽게 알 수 있습니다. 
하지만 적색 영역에 너무 많은 Routing path가 집중되어 있어 실제 신호 전달 시에는 녹색 선으로 우회해서 신호가 전달됩니다. 
이러한 예시를 통해 우리의 목표를 아래와 같이 정의할 수 있습니다.

*<center> '소자들의 거리(WireLength)는 최대한 가깝게 하되, </center>*
*<center> 한곳에 너무 많은 Routing path(Congestion)들이 몰리지 않게 !'  </center>*

위의 목표를 생각해 봤을 때, 소자들의 위치는 최대한 가까워야 하지만, 또 너무 근접하면 안 되는, **Multi-objective Optimization Problem** 이라는 것을 알 수 있습니다. 
이를 달성하기 위해 설계한 보상함수는 아래와 같습니다.

$$
R_{p,g}=-\alpha \cdot WireLength(p,g) - \beta \cdot Congestion(p,g) \qquad \text{S.t. } \text{density}(p,g) \leq \text{max}_\text{density}
$$

WireLength는 소자들을 가깝게 배치해주기 위한 역할, Congestion은 Routing Path들이 한곳에 너무 집중되지 않게 하기 위한 역할을 합니다. 
아래 내용에서 각 항목이 어떻게 계산되고 적용되는지 설명합니다.

> ### WireLength

Reviewer: Class name이 아니므로 띄어쓰기를 쓰면 어떨까요?
```suggestion
### Wire Length
```

WireLength를 근사하기 위해 Half-Perimeter Wire Length(HPWL)를 사용합니다. 
모든 Net에 대하여, 각 Net에 속해있는 소자들에 대해 경계 상자를 그린 후, 그 둘레의 절반을 Wirelength로 근사하는 것입니다. 
HPWL은 가장 일반적으로 사용되는 WireLength 추정 모델입니다.

> 계산이 매우 간단하고 소자의 Pin 수가 2 또는 3인 경우의 Net에 대해서 직교 Steiner Tree 모델의 길이와 정확히 일치한다는 장점이 있습니다.

Reviewer: '직교 Steiner Tree 모델'이 무엇인가요?

수식으로 표현하면 아래와 같습니다.

$$
\begin{aligned}

> HPWL(i) = (max_{b\in i}\{x_b\} - min_{b\in i}\{x_b\} + 1) \\ + (max_{b\in i}\{y_b\} - min_{b\in i}\{y_b\} + 1)

Reviewer: ```suggestion
HPWL(i) = (\max_{b\in i}\{x_b\} - \min_{b\in i}\{x_b\} + 1) \\ + (\max_{b\in i}\{y_b\} - \min_{b\in i}\{y_b\} + 1)
```

\end{aligned}
$$

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/wirelength.png" alt="hpwl calculation" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - HPWL 계산 </figcaption>
</p>
</figure>

왼쪽의 그림에서 Routing path의 색깔이 각기 다른 Wire를 나타낸다고 했을 때, HPWL은 아래와 같이 계산됩니다.

$$
HPWL(Red)=(2 - 1 + 1) + (2 - 1 + 1) = 4 \\
HPWL(Green)=(4 - 1 + 1) + (1 - 1 + 1) = 5 \\

> HPWL(Yellow)=(4 - 1 + 1) + (4 - 1 + 1) = 8

Reviewer: ```suggestion
HPWL(Yellow)=(3 - 1 + 1) + (4 - 1 + 1) = 7
```

Author: 큰일날 뻔 했네요 ㅋㅋㅋㅋ 감사합니다.

$$

> 이렇게 전체 WireLength의 근사치는 4 + 5 + 8 = 17이 됩니다. 

Reviewer: ```suggestion
이렇게 전체 WireLength의 근사치는 4 + 5 + 7 = 16이 됩니다. 
```

이러한 계산방식으로 Net들의 WireLength를 줄입니다.

###### Example Code
```python
def calculate_hpwl(net: Net) -> float:
    """Calculate HPWL given Net."""
    min_row, min_col, max_row, max_col = get_boundary_coordinate_of_net(net)
    hpwl = (max_row - min_row + 1.0) + (max_col - min_col + 1.0)
    return hpwl
```

## Congestion
Congestion은 Routing Path들이 특정 지역에만 모이지 않게 하기 위해 적용됩니다.
저희가 환경 설계 과정에서 굉장히 많이 고민했던 부분이기도 합니다.
Congestion은 특정 구역에서 제공되는 Routing Resources (supply)와 해당 위치에 놓인 Net의 Routing Path (demand)로 정의됩니다[[7](#ref-7)].
Horizontal과 Vertical의 Routing Resources가 독립적으로 사용되기 때문에 Congestion 계산 시에도 따로 고려합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/congestion_def.png" alt="cogestion definition" width="100%">

>   <figcaption style="text-align: center;">[그림 <script>func()</script>] - Congestion 정의 </figcaption>

Reviewer: 그림에 출처표기가 필요해 보입니다.
```suggestion
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Congestion 정의 [7]</figcaption>
```

</p>
</figure>
<!--https://beckassets.blob.core.windows.net/product/readingsample/497806/9780387300375_excerpt_001.pdf-->

> 위의 정의를 고려하여, 저희가 설계한 Congestion 계산 방식은 아래와 같습니다.

Reviewer: 저희는 supply는 계산에 이용하지 않음을 언급할 필요가 있습니다.
(demand가 높아지거나 낮아지면 congestion도 높아지거나 낮아짐. supply는 고정값)

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/congestion_calculation.png" alt="congestion calculation flow" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Congestion 계산 과정 </figcaption>
</p>
</figure>

###### Example Code

> ```python
> def accumulate_local_congestion(
> 	  global_congestion_map: np.ndarry,
> 	  net: Net
> 	) -> None:
> 	    """Accumulate local congestions into the global congestion map."""
> 	    # get edges through MST or Steiner Tree
> 	    edges = get_edges(net)
> 	
> 	    for edge in edges:
> 	        # Calculate all path number to source / target for all grid cells.
> 	        paths = get_paths(two_pnt_net)
> 	        # Calculate partial congestion map horizontalㅣy and verticalㅣy
> 	        congestion_maps = get_congestion_map(paths)
> 	    global_congestion_map += congestion_maps
> ```

Reviewer: Invalid indentation

```suggestion
```python
def accumulate_local_congestion(
    global_congestion_map: np.ndarry,
    net: Net
) -> None:
    """Accumulate local congestions into the global congestion map."""
    # get edges through MST or Steiner Tree
    edges = get_edges(net)
	
    for edge in edges:
	# Calculate all path number to source / target for all grid cells.
	paths = get_paths(two_pnt_net)
	# Calculate partial congestion map horizontalㅣy and verticalㅣy
	congestion_maps = get_congestion_map(paths)
    global_congestion_map += congestion_maps
```
```

</p>
</figure>

GNN은 그래프 형태로 표현된 Netlist 정보를 Embedding 해주기 위해 사용했습니다.
Graph Feature embedding은 아래의 세부 단계를 거치며 수행됩니다.

1. Macro Features와 Netlist Graph를 GNN에 입력하여 Node Embedding을 추출
2. Placement Map에 각 소자들의 배치 현황을 표시
3. Placement Map의 정보를 바탕으로 Node Embedding을 Aggregation
4. 3의 결과물의 채널에 배치할 Macro Embedding을 브로드캐스팅하여 Concatenation

즉, 각 Bin에는 배치된 소자들의 Aggregated Embedding(현재 배치상태 + Netlist 연결정보)과 새로 배치할 Macro Embedding(현재 배치할 소자 정보)이 포함됩니다.
3차원의 Embedding 정보를 사용하면 CNN을 활용해 위치에 따른 지역 정보를 효과적으로 인식할 수 있다는 장점이 있고 이는 결국 적절한 배치영역을 찾을 수 있도록 정보를 제공합니다.

## Action Decision
Graph Feature Embedding 된 정보를 하나의 이미지로써 인식해 다음 소자가 배치될 확률분포를 계산합니다. 
마지막 Layer의 계산 결과에 Invalid Action Masking*이 적용되며, 해당 값들을 이용해 소자의 위치를 결정합니다.
이 과정은 아래 그림처럼 표현됩니다.

\*  <font size="-1"> Invalid Action Masking: 소자들이 놓인 상태에 따라 배치 가능한 위치를 결정해주는 작업. (APPENDIX A.2 참고). </font>

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/action_decision.png" alt="policy network" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Action Decision </figcaption>
</p>
</figure>

Graph Feature Embedding을 CNN에 입력하여 점진적으로 채널을 1까지 줄입니다. 

> Input Channel이 점진적으로 줄어드는 과정에서 Receptive Field는 커지므로, 결국에는 전 영역을 모두 볼 수 있게끔 조절할 수 있습니다. 

Reviewer: 무엇을 조절할 수 있나요? (e.g. 깊이)

마지막 Layer의 출력은 현재 소자를 배치할 위치에 대한 확률분포를 나타냅니다. 
에이전트는 해당 확률분포에 따라 행동을 결정합니다.

# Experiment Setup & Details
### Dataset
강화학습의 효용성을 검증하기 위해 아래 Netlist를 사용했습니다.

> 상세 스펙은 아래와 같습니다(by Auto Placement of ICC2).

Reviewer: ```suggestion
상세 스펙은 아래와 같습니다(그림의 배치는 ICC2의 Macro Auto Placement 결과).
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/dataset.png" alt="dataset specification" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Dataset specification</figcaption>
</p>
</figure>

해당 Netlist는 18개의 Macro와 약 12만 개의 Standard Cell로 구성되어 있습니다.
Standard Cell들을 50개의 Cluster로 압축한 후, 아래 기술된 실험을 진행했습니다.

> Graph Feature Embedding을 위해 Graph Convoltion Network(GCN) 모델을 사용했습니다.

Reviewer: ```suggestion
Graph Feature Embedding을 위해 Graph Convoltional Network(GCN) 모델을 사용했습니다.
```

    <td> 2.60 </td> <td> 227 </td>
    <td> 2.55 </td> <td style="color:red"> <strong> 230 </strong> </td>
  </tr>
  <tr align="center">
    <td bgcolor="Silver"> Group 5 </td>
    <td> 2.67 </td> <td> 231 </td>
    <td style="color:red"> <strong> 2.74 </strong> </td> <td style="color:red"> <strong> 235 </strong> </td> 
    <td> 2.73 </td> <td> 234 </td>
    <td> 2.68 </td> <td> 231 </td>
    <td> 2.63 </td> <td> 234 </td>
  </tr>
</table>

Floorplan 단계에서 가장 주의 깊게 봐야 할 수치는 WNS, FREQ 가 있습니다.
WNS는 신호가 각 소자에 전달될 때 설계 시 예측했던 시간과 실제 도착시간의 차이를 의미합니다.
만약 0보다 작으면 설계된 조건에 대한 Violation으로 취급되며 Floorplan을 다시 진행해야 합니다.
FREQ는 초당 처리한 수 있는 신호의 수, 즉 연산속도를 의미합니다.
그러므로, WNS는 양수인지를 확인하고, FREQ는 값을 비교합니다.
각 그룹은 성능 검사를 위해 미리 정해둔 Routing Path의 집합입니다.
결과적으로 **'각 그룹에 대해 WNS가 0 이상이면서 FREQ가 높을 때 성능이 좋다'**라고 말할 수 있겠습니다. (각 지표에 대한 자세한 설명은 @jiyoon.lim 포스팅을 참고해주세요.)
결과를 보시면, 저희의 에이전트는 두 가지 Baseline에 근접한 수치를 보여줍니다.

# 앞으로 풀어야 할 숙제는?
MakinaRocks COP팀에서 개발 중인 강화학습 기반의 알고리즘은 약 12만 개 Standard Cell로 구성된 비교적 작은 규모의 Netlist에 대해서 유의미한 학습 결과를 보였습니다.
이를 기반으로 더 큰 규모의 Netlist에 대한 학습할 예정이며 궁극적으로는 Generalize를 실현해보려 합니다.
목표를 달성하기 위해 앞으로 해결해야 할 문제들이 남아있습니다.

> * 더 적은 학습 시간과 더 좋은 배치 결과를 얻을 방법은?

Reviewer: ```suggestion
* 더 적은 시간의 학습으로부터 더 좋은 배치 결과를 얻을 수 있는 방법은?
```

5. State (Clustering, Feature 정보)
6. Action (Grid Cell이라는 이산공간 정의, 셀의 배치순서)
7. Reward에 대한 설명 (HPWL / Congestion)
8. 그렇다면 State로부터 Action을 어떻게 도출하나? (Model - GNN + CNN): Preprocessing을 여기서 언급
9. 실험 세팅 (강화학습 Agent / 장비 / ...)
10. 실험 결과
11. Further Works 
-->

<script>
    var figure_idx = 1
    function func()
    {
        document.write(figure_idx)
        figure_idx = figure_idx + 1
    }
</script>

# Related to
* [강화학습, 산업의 난제에 도전하다! - ASIC 반도체 설계 (Floorplan) 자동화 (Naver Deview 21 발표 영상)](https://tv.naver.com/v/23650123) 
* [포스팅 1부](@jiyoon)

# 프로젝트 소개

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/placements.gif" alt="placement animation" width="100%">
  <figcaption style="text-align: center;">
  <p style="font-weight: bold;"> FIGURE ABSTRACT </p>
  위 애니메이션은 약 1500 iterations 정도의 학습 과정을 보여주고 있습니다.
  녹색의 큰 사각형은 Macro, 청색의 작은 사각형은 Standard Cell Cluster를 나타냅니다. 
  어느 정도 학습이 진행된 후, 사각형의 움직임이 눈에 띄게 줄어들며 안정된 모습을 보여줍니다.
  에이전트는 Macro는 Chip Canvas 영역의 바깥쪽, Standard Cell Cluster는 안쪽에 위치시키도록 학습합니다.
  이에 대한 자세한 설명은 본 포스팅을 참고해주세요.
  </figcaption>
</p>
</figure>

----
<p></p>

#### Introduction

> 본 프로젝트는 지난 2020년 6월 구글에서 발표해 Nature 등재로 화제가 되었던 **A graph placement methodology for fast chip design**[[1](#ref-1)][[2](#ref-2)] 논문에 기초를 두고 있습니다.
> 강화학습을 이용해 주문형 반도체 (Application-Specific Integrated Circuit, ASIC)의 제조 공정 중 하나인 Floorplan 단계에서의 Optimization 문제를 해결하고자 합니다.

Reviewer: Minor한 제안입니다. 참고만 해주세요.

목적 -> 배경 -> 글에서 소개할 내용 언급

```suggestion
본 프로젝트의 목적은 강화학습을 이용해 주문형 반도체 (Application-Specific Integrated Circuit, ASIC)의 제조 공정 중 하나인 Floorplan 단계에서의 Optimization 문제를 해결하는 것입니다. 
주요 기술 내용은 지난 2020년 6월 구글에서 발표해 Nature 등재로 화제가 되었던 **A graph placement methodology for fast chip design**[[1](#ref-1)][[2](#ref-2)] 논문에 기초를 두고 있습니다.
본 포스팅을 통해 약 5개월간 경험한 Floorplan 문제의 이해부터 솔루션 구현까지의 과정을 공유드리려고 합니다.

# Floorplan
Floorplan은 트랜지스터가 아닌 블록 단위의 배치를 하는 과정이며, 수십~수백만 개 단위의 트랜지스터와의 연결성을 고려하며 블록들을 배치해야 하기 때문에 시간이 많이 소요되는 작업 중 하나입니다.
...
```

Author: Floorplan은 @jiyoonlim123 님 포스팅에서 언급하므로 따로 챕터를 만들지 않는게 좋을거같은데 어떻게 생각하세요? 지윤님 포스팅을 레퍼런싱 해도 좋을거같습니다.

Reviewer: `Floorplan 문제의 이해부터 솔루션 구현까지의 과정을 공유드리려고 합니다.`
라고 위에 서술한다면 간단히라도 언급하는 것이 좋겠습니다.

위 문장을 조금 수정하셔도 좋구요.

Reviewer: 문맥을 다시 보니 수정하지 않아도 괜찮을 것 같습니다.
Resolve 하겠습니다.

> Floorplan은 소자들을 배치하는 과정 중 하나이며, 트랜지스터가 아닌 블록 단위의 배치를 말합니다.
> 수십~수백만 개 단위의 트랜지스터와의 연결성을 고려하며 블록들을 배치해야 하기 때문에 시간이 많이 소요되는 작업 중 하나입니다.

Reviewer: 내용상 연결성이 강한 두 문장 연결.
(분리하여 적는 것이 기술적으로 어려울 때는 합치는 것이 쉽습니다.)

```suggestion
Floorplan은 트랜지스터가 아닌 블록 단위의 배치를 하는 과정이며,
수십~수백만 개 단위의 트랜지스터와의 연결성을 고려하며 블록들을 배치해야 하기 때문에 시간이 많이 소요되는 작업 중 하나입니다.
```

본 포스팅에서는 약 5개월간 경험한 Floorplan 문제의 이해부터 솔루션을 구현까지의 과정을 공유드리겠습니다.

> 먼저 본 포스팅에서 자주 언급되는 주요 용어들과 뜻입니다.

Reviewer: `본 포스팅` 중복사용
```suggestion
먼저 본문에서 자주 언급하는 주요 용어들과 뜻입니다.
```

* Macro: Floorplan 과정에서 배치되는 소자의 종류 중 하나, 메모리 구성 요소로 Standard Cell 대비 훨씬 큰 크기를 가지고 있음.
* Standard Cell: NAND, NOR, XOR과 같은 논리 게이트, 크기가 매우 작으며 수십~수백만 개 단위의 소자가 사용됨.
* Net: Macro와 Standard Cell들의 연결 관계를 정의하는 Wire.
* Netlist: Net들의 집합체.

# Is that Real Problem?
Floorplan 과정은 일반적으로 사람의 손에 의해 수행됩니다.
Chip Canvas라는 2차원의 연속 공간에서 복잡한 연결 관계를 고려하며 수백 개에 달하는 Macro들을 배치하는 작업은 큰 탐색공간(Search Space)으로 인해 높은 복잡도를 가지고 있는 문제입니다.
또한, 직접 Macro를 배치한 후 그 뒤의 설계과정을 상용 Electronic Design Automation(EDA) Tool을 통해 수행하고 나서야 배치에 대한 평가를 할 수 있습니다.
만약 평가 결과가 좋지 않다면, Macro를 배치하는 작업부터 다시 수행해야 하므로 숙련된 기술자라도 며칠에서 몇 주의 시간을 소요하기도 합니다.
반복적인 경험을 통해 습득한 지식을 기반으로 진행되는 Floorplan은 기술자의 숙련도에 따라서 필요한 시간이 크게 달라지곤 합니다.
현업에서 종사하시는 전문가분들과의 협업을 통해 알게 된 Floorplan의 주요 문제점은 아래와 같이 정리할 수 있습니다.

* Macro의 배치를 수작업으로 진행 후, EDA Tool을 이용해 결과를 확인하는 작업을 여러 번 반복해야 하는 만큼 시간이 오래 걸림.
* 사람의 경험이 매우 중요하기 때문에, 숙련도에 따라 품질과 요구되는 시간의 편차가 큼.

> 연구 배경 및 의의에 대한 더욱 자세한 설명은 관련 포스팅을 참고해주세요. (@jiyoon.lim)
> 이 포스팅에서는 문제정의와 해결 방법 관련된 기술적인 내용을 중점적으로 설명합니다.

Reviewer: 문장 순서를 바꾸는 것이 더 자연스러워 보입니다.
(중요한 것을 좀 더 앞에)
```suggestion
이 포스팅에서는 문제정의와 해결 방법 관련된 기술적인 내용을 중점적으로 설명합니다.
(연구 배경 및 의의에 대한 더욱 자세한 설명은 [관련 포스팅](@jiyoon.lim)을 참고해주세요.)
```

# How to solve?
기존의 배치 최적화 문제*에 적용되는 많은 Classic한 방법들(Simulated Annealing[[3](#ref-3)], Genetic Algorithm[[4](#ref-4)]...)은 좋은 성능을 보여주고 있지만, 새로운 문제를 만나면 같은 도메인이라도 초기 상태부터 다시 최적화를 진행해야 한다는 단점이 있습니다.

> 한번 학습된 모델이 다르게 정의된 문제에서 별다른 추가 학습 없이 어느 수준 이상의 해답을 찾는 것을 일반화(Generalize)라고 합니다.
> 소자 5개, 10개로 구성된 Netlist를 학습하여 20개 규모의 Netlist에 대해서도 일정 수준 이상의 결과를 얻어내는 것이 그 예시입니다.
> 이미 많은 연구에서 신경망을 직접 시킨 강화학습은 강력한 Generalize 수준을 보여주고 있습니다.
> 강화학습의 이러한 강점을 이용한다면 **여러 구조와 크기의 Netlist에 대해서 추가적인 학습 없이 또는 최소한의 학습 시간만**을 가지고 좋은 배치를 찾을 수 있습니다.
> 저희는 이점에 주목하여 강화학습을 주요 알고리즘으로써 사용했습니다.

Reviewer: - 문장간의 연결이 더 부드러워지면 좋겠습니다. 애매할 때는 문장 순서를 바꿔보는 것도 좋습니다.
- '이미 많은 연구'는 레퍼런싱이 필요해보입니다. (e.g. PointerNet, Structure2Vec)

```suggestion
한번 학습된 모델이 다르게 정의된 문제에서 별다른 추가 학습 없이 어느 수준 이상의 해답을 찾는 것을 일반화(Generalize)라고 합니다.
이미 많은 연구에서 인공 신경망을 적용한 강화학습은 강력한 Generalize 수준을 보여주고 있습니다.
강화학습의 이러한 강점을 이용한다면 **여러 구조와 크기의 Netlist에 대해서 추가적인 학습 없이 또는 최소한의 학습 시간만**을 가지고 좋은 배치를 찾을 수 있습니다.
(예: 소자 5개, 10개로 구성된 Netlist 들의 배치를 학습하여 20개 규모의 Netlist에 대해서도 일정 수준 이상으로 배치)
저희는 이점에 주목하여 강화학습을 주요 알고리즘으로써 사용했습니다.
```

> 사람이 반복적인 경험을 통해 직관적으로 수행하는 업무를 데이터 기반으로 학습하여 해결하는 것의 최종목표입니다.
> 본 포스팅에는 최종 목표에 도달하기에 앞서, 강화학습을 사용한 단일 Netlist의 Floorplan 과정을 소개합니다.
> 아래에서는 먼저 강화학습이 무엇인지에 대해 간단히 기술하고 저희가 이 문제에 대해 어떻게 접근하고 있는지 설명합니다.

Reviewer: 개인적으로는 맨 앞, 맨 뒤 문장이 없는 것이 더 자연스러워 보입니다.
부자연스럽게 보인 부분은 아래와 같습니다.
- 바로 직전까지 Generalization을 기술 -> 이번 문장에서는 데이터 기반 학습이 최종목표라 언급
- 본 포스팅에서는... 에서 이미 뒤이어 나올 내용을 소개.
```suggestion
본 포스팅에는 최종 목표(Gneneralization)에 도달하기에 앞서, 강화학습을 사용한 단일 Netlist의 Floorplan 학습 과정을 소개합니다.
```

\*  <font size="-1"> 배치 최적화 문제: 대상의 물리적인 위치를 통해 목표함수를 최적화하는 문제 </font>

### 강화학습이란?
강화학습은 에이전트(Agent)가 환경(Environment)과 상호작용에서의 시행착오(Trial and Error)를 통해 더 높은 보상(Reward)을 얻을 수 있는 행동(Action)들을 학습하는 알고리즘입니다.
환경은 에이전트의 행동마다 다음 상태(State)와 보상을 반환하며, 에이전트는 전달받은 상태를 고려해 다시 행동을 결정하게 됩니다.
강화학습은 **Markov Decision Process**[[5](#ref-5)] (MDP)에 기반해, 차례로 어떤 의사결정(Decision Process)을 해야 하는 문제들에 많이 적용됩니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/mdp.png" alt="markov decision process" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Markov Decision Process</figcaption>
</p>

> </figure>

Reviewer: GNN의 입력을 지우기보다 환경에서의 입력이 어디에 들어가는지 추가로 표시해주는 것이 이해에 도움이 될 것 같습니다.

결국 에이전트는 학습하는 과정 중에 Sampling 되는 모든 에피소드(Episode)*에 대한 보상의 기댓값을 최대화하기 위한 행동들을 학습합니다.
당연하게도, 강화학습은 그 구성 요소들(에이전트, 상태, 행동, 보상)의 정의방법에 따라 다른 학습양상을 보여줍니다.
아래에서는 저희가 환경과 에이전트를 어떻게 설계했는지 설명드리겠습니다.

\*  <font size="-1"> Episode: 에이전트가 일련의 행동들을 수행하고 보상들을 얻는 Sequence, 에피소드가 종료되면 환경은 초기상태로 리셋된다. </font>

### Technical Overview
각 단계의 자세한 내용을 설명해 드리기 앞서 환경을 어떻게 정의하고 모델을 어떻게 설계했는지에 대해 전체적인 흐름을 설명합니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/technical_overview.png" alt="technical overview" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Technical Overview</figcaption>
</p>
</figure>

##### 환경 정의
에이전트는 환경의 주요 구성요소인 상태, 행동, 보상을 통해 환경과 상호작용하며 학습합니다.
그러므로 에이전트를 효과적으로 학습시키기 위해서는 각 요소를 적절히 정의해야 합니다.
저희의 경우 각 요소는 아래와 같은 역할을 수행해야 합니다.

* 상태: 행동을 결정하기 위한 판단에 필요한 정보를 담고 있어야 함.
* 행동: 더 좋은 Chip Design을 얻을 수 있게 하는 행동의 집합.
* 보상: Chip Design의 품질.

저희가 설계한 환경은 아래와 같은 순서로 진행됩니다.

1. 빈 Chip Canvas인 초기 상태에서 시작하여 행동 한 번에 소자 한 개를 배치.
2. 마지막 배치 외의 모든 배치에 대해 0의 보상을 반환.
3. 1~2번 반복 후, 모든 소자들의 배치가 완료되면 보상을 계산.
4. 1~3번 반복하며 학습 진행.

> 상태, 행동, 보상 정의에 대한 자세한 설명은 아랫부분을 참고해주세요.

Reviewer: ```suggestion
상태, 행동, 보상 정의에 대한 자세한 설명에 앞서 모델 설계를 먼저 알아보도록 하겠습니다.
```

##### 모델 설계

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/overview_model.png" alt="role op model" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - 모델의 역할 </figcaption>
</p>
</figure>

Neural Network Model은 환경으로부터 전달받은 상태의 정보들을 인식해 행동을 결정할 확률분포를 계산하는 역할을 수행합니다.
이 프로젝트의 경우, 소자의 물리적 특성과 연결정보를 바탕으로 이미 배치된 소자들에 대한 정보와 앞으로 배치할 소자들에 대한 정보를 알아야 합니다.
이를 구현하기 위해, Graph Neural Network(GNN)과 Convolutional Neural Network(CNN)으로 이루어진 모델을 설계했습니다.
각각의 목적은 아래와 같습니다.

* GNN: Graph 형태의 Netlist의 정보를 효과적으로 Embedding하기 위해 선택했습니다.
* CNN: 배치가 완료된 소자들의 지역적인 정보를 효과적으로 학습할 수 있는 CNN 구조를 선택했습니다.

> 이에 대한 자세한 설명 역시 아래에서 설명합니다.

Reviewer: 좀 더 구체적으로 기술하면 어떨까요?
```suggestion
모델의 각 구조에 대한 자세한 설명은 "Model Description" 장에서 드리도록 하겠습니다.
```

# 강화학습 환경 정의하기
## State Representation
상태는 현재 환경이 어떤 상태에 있는지에 대한 정보를 담고 있습니다.
그러므로, 상태는 에이전트가 좋은 행동을 결정하기 위해 도움이 되는 정보들을 전달해야 합니다.
칩 설계의 전문가분들이 **소자들의 연결 관계**와 **물리적 특성**들을 고려해 배치하는 것에 착안하여 에이전트가 다음 행동을 결정하기 위해 필요한 정보를 아래와 같이 정의했습니다.

1. 소자들 간의 연결 관계
2. 각 소자의 물리적 정보 (width, height, is_macro, n_pins ... etc.)
3. 이번에 배치할 소자의 정보
4. 지금까지의 배치 정보

각 소자들 간의 연결 관계가 아래와 같은 Hypergraph*의 형태를 띠고 있는 만큼, 소자들의 물리적 특성들을 Node Features로 정의할 수 있습니다. 

\*  <font size="-1"> Hypergraph: Edge한개에 다수개의 Vertex가 연결되어 있는 그래프 구조 </font>

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/hypergraph.png" alt="hypergraph" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Hypergraph 예시 </figcaption>
</p>

> </figure>

Reviewer: 그림 출처 표시가 필요합니다.

Graph Clustring Alogrithm을 이용하여 수십~수백만 개 단위의 Standard Cell들에 Clustering을 진행합니다.
결과적으로, 배치하려는 대상의 개수가 줄어들어 문제의 복잡도가 감소합니다.
Clustering 과정에서의 정보손실을 최소화하기 위해 Hypergraph Clustering Algorithm인 hMETIS[[6](#ref-6)] 알고리즘을 사용합니다. 

> 추후 배치 과정에서 한 Cluster에 속해있는 소자들을 묶어 하나의 독자적인 소자로 취급합니다.

Reviewer: 문장들이 약간 따로 움직이는 느낌입니다. 
위에 드린 코멘트를 참고하셔서 문장 연결을 다듬어보시면 어떨까 싶습니다.

## Action Representation
행동은 에이전트가 상태의 정보들(소자들의 물리적 특성과 현재 배치상태 등)을 이용해 현재 배치할 소자를 어느 위치에 둘 것인지 결정짓는 역할을 합니다. 
에이전트가 얻는 보상을 최대화하기 위해서는 충분한 탐험(Exploration, 경험)이 필요합니다.
경험을 효율적으로 하기 위해서는 적절한 크기의 탐색공간을 정의해야 합니다.
저희는 이러한 점들을 고려해 연속공간의 Chip Canvas를 2차원 Grid의 이산 공간으로 정의했습니다.
에이전트가 Neural Network로부터 얻어진 확률분포에서 특정 Grid Cell을 선택하면 배치될 소자의 중앙이 해당 Grid Cell의 중앙에 위치하도록 구현했습니다.

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/action.png" alt="action" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Action </figcaption>
</p>
</figure>

저희가 구현한 방식은 탐색공간을 줄이면서 배치 결과를 평가하기 위한 Approximated Metric을 단순하게 정의할 수 있다는 장점이 있습니다.
해당 단계에서 고려되어야 할 아래 issue들에 대해 아래 [APPENDIX](#appendix)에 정리했습니다.

* [A.1. Grid Cell Size Decision](#a.1.)
* [A.2. Invalid Action Masking](#a.2.)

## Reward Representation
보상은 에이전트에게 해당 행동이 얼마나 유익한지를 알려주는 학습에 굉장히 중요한 요소입니다. 
일반적으로 ASIC의 Floorplan 결과는 ICC2와 같은 상용 EDA Tool을 사용하여 측정됩니다. 

> Floorplan 이후의 배치 결과를 평가하기 위한 추가 설계과정은 모두 해당 Tool에서 수행되기 때문에, 보통 수 시간이 소요됩니다.

Reviewer: `Tool에서 수행되기 때문에 -> 수시간이 소요된다.`의 인과관계가 독자에게 자연스럽게 받아들여지기 어려울 수 있을 것 같습니다.

```suggestion
Floorplan 이후의 배치 결과를 평가하기 위해 EDA Tool에서 수행하는 추가 설계과정은 보통 수 시간이 소요됩니다.
```

  hpwl = (max_row - min_row + 1.0) + (max_col - min_col + 1.0)
  return hpwl
```

## Congestion
Congestion은 Routing Path들이 특정 지역에만 모이지 않게 하기 위해 적용됩니다.
저희가 환경 설계 과정에서 굉장히 많이 고민했던 부분이기도 합니다.
Congestion은 특정 구역에서 제공되는 Routing Resources (supply)와 해당 위치에 놓인 Net의 Routing Path (demand)로 정의됩니다[[7](#ref-7)].
Horizontal과 Vertical의 Routing Resources가 독립적으로 사용되기 때문에 Congestion 계산 시에도 따로 고려합니다.

<p align = "center">
  <img src = "/assets/images/2021-11-17-Deview-details/congestion_def.png">
</p>
<p align = "center">
[그림 <script>func()</script>] - Congestion 정의[<a href="https://beckassets.blob.core.windows.net/product/readingsample/497806/9780387300375_excerpt_001.pdf">7</a>]
</p>
<!-- [[7](#ref-7)] -->
<!-- <figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/congestion_def.png" alt="cogestion definition" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Congestion 정의 <ref=ref-7> </figcaption>
</p>
</figure> -->
<!--https://beckassets.blob.core.windows.net/product/readingsample/497806/9780387300375_excerpt_001.pdf-->

위의 정의에서 Supply는 고정값을 가지며 Demand에 따라서 Congestion이 달라집니다.

> 이러한 점을 고려해, 저희가 설계한 Congestion 계산 방식은 아래와 같습니다.

Reviewer: ```suggestion
이러한 점을 고려해, 저희는 아래와 같이 계산한 Demand의 추정치를 Congestion 값으로 사용합니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/congestion_calculation.png" alt="congestion calculation flow" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Congestion 계산 과정 </figcaption>
</p>

> </figure>

Reviewer: Suggestion:
5. (0,1) Grid Cell에 대한 Expected Routing Resource Demand를 계산
6. 5와 같은 방식으로 다른 Grid Cell에 대해 얻은 계산결과를 Global Congestion Map에 반영

Reviewer: 받아들이는 입장에서 이해하기 가장 어려워하는 부분 중 하나입니다.
특히 신경써서 설명을 친절하게 하면 좋겠습니다.

###### Example Code
```python
def accumulate_local_congestion(
  global_congestion_map: np.ndarry,
  net: Net
) -> None:
  """Accumulate local congestions into the global congestion map."""
  # get edges through MST or Steiner Tree
  edges = get_edges(net)

  for edge in edges:
    # Calculate all path number to source / target for all grid cells.
    paths = get_paths(two_pnt_net)

>     # Calculate partial congestion map horizontalㅣy and verticalㅣy

Reviewer: ```suggestion
    # Calculate partial congestion map horizontally and vertically
```

  <img src="/assets/images/2021-11-17-Deview-details/h_to_o.png" alt="hypergraph to ordinary graph" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Hypergraph to Ordinary Graph</figcaption>
</p>
</figure> 

## Graph Features Embedding
<!-- 추후 구현할 Generalize를 위해서 동일한 GNN모델을 사용해 다양한 구조의 Netlist들의 특성들을 인식할 수 있어야 합니다.
또한, 큰 그래프를 인식하기 위해 GNN의 Layer수가 늘어나면서 노드의 임베딩이 서로 유사해지는 문제인 Oversmoothing 문제도 고려해야합니다.
저희가 사용한 DeeperGCN[ ]의 경우 Residual Connecteion을 사용해 GNN의 고질적인 단점이였던 Oversmoothing 문제를 효과적으로 억제합니다. 
Computer Vision 분야에서 좋은 성능으로 많은 사랑을 받았던 ResNet[ ]처럼 DeeperGCN도 동일한 크기의 모델로 다양한 구조와 크기의 그래프에 대해 전체적으로 좋은 성능을 보여주었습니다. 
DeeperGCN을 통한 Feature Embedding의 과정은 아래 그림에서 보실 수 있습니다. -->

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/graph_embedding.png" alt="Graph Embedding" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Graph Feature Embedding</figcaption>
</p>
</figure>

GNN은 그래프 형태로 표현된 Netlist 정보를 Embedding 해주기 위해 사용했습니다.
Graph Feature embedding은 아래의 세부 단계를 거치며 수행됩니다.

1. Macro Features와 Netlist Graph를 GNN에 입력하여 Node Embedding을 추출

> 2. Placement Map에 각 소자들의 배치 현황을 표시

Reviewer: ```suggestion
2. Placement Map에 각 소자들의 배치 현황을 표시 (각 row, col에 배치된 cell index의 값을 1로 설정, otherwise 0)
```

    <td style="color:red"> <strong> 0.79 </strong> </td> <td style="color:red"> <strong> 161 </strong> </td>
    <td> 0.65 </td> <td> 158 </td>
    <td> 0.69 </td> <td> 158 </td>
    <td style="color:red"> <strong> 0.79 </strong> </td> <td> 160 </td>
  </tr>
  <tr align="center">
    <td bgcolor="Silver"> Group 4 </td>
    <td> 2.60 </td> <td> 227 </td>
    <td> 2.65 </td> <td style="color:red"> <strong> 230 </strong> </td>
    <td style="color:red"> <strong> 2.66 </strong> </td> <td style="color:red"> <strong> 230 </strong> </td>
    <td> 2.60 </td> <td> 227 </td>
    <td> 2.55 </td> <td style="color:red"> <strong> 230 </strong> </td>
  </tr>
  <tr align="center">
    <td bgcolor="Silver"> Group 5 </td>
    <td> 2.67 </td> <td> 231 </td>
    <td style="color:red"> <strong> 2.74 </strong> </td> <td style="color:red"> <strong> 235 </strong> </td> 
    <td> 2.73 </td> <td> 234 </td>
    <td> 2.68 </td> <td> 231 </td>
    <td> 2.63 </td> <td> 234 </td>
  </tr>
</table>

Floorplan 단계에서 가장 주의 깊게 봐야 할 수치는 WNS, FREQ 가 있습니다.
WNS는 신호가 각 소자에 전달될 때 설계 시 예측했던 시간과 실제 도착시간의 차이를 의미합니다.
만약 0보다 작으면 설계된 조건에 대한 Violation으로 취급되며 Floorplan을 다시 진행해야 합니다.
FREQ는 초당 처리한 수 있는 신호의 수, 즉 연산속도를 의미합니다.

> 그러므로, WNS는 양수인지를 확인하고, FREQ는 값을 비교합니다.

Reviewer: ```suggestion
그러므로, WNS는 양수인지를 확인하고, FREQ는 값의 크기를 비교합니다.
```

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/h_to_o.png" alt="hypergraph to ordinary graph" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Hypergraph to Ordinary Graph</figcaption>
</p>
</figure> 

## Graph Features Embedding
<!-- 추후 구현할 Generalize를 위해서 동일한 GNN모델을 사용해 다양한 구조의 Netlist들의 특성들을 인식할 수 있어야 합니다.
또한, 큰 그래프를 인식하기 위해 GNN의 Layer수가 늘어나면서 노드의 임베딩이 서로 유사해지는 문제인 Oversmoothing 문제도 고려해야합니다.
저희가 사용한 DeeperGCN[ ]의 경우 Residual Connecteion을 사용해 GNN의 고질적인 단점이였던 Oversmoothing 문제를 효과적으로 억제합니다. 
Computer Vision 분야에서 좋은 성능으로 많은 사랑을 받았던 ResNet[ ]처럼 DeeperGCN도 동일한 크기의 모델로 다양한 구조와 크기의 그래프에 대해 전체적으로 좋은 성능을 보여주었습니다. 
DeeperGCN을 통한 Feature Embedding의 과정은 아래 그림에서 보실 수 있습니다. -->

<figure class="image" style="align: center;">
<p align="center">
  <img src="/assets/images/2021-11-17-Deview-details/graph_embedding.png" alt="Graph Embedding" width="100%">
  <figcaption style="text-align: center;">[그림 <script>func()</script>] - Graph Feature Embedding</figcaption>
</p>
</figure>

GNN은 그래프 형태로 표현된 Netlist 정보를 Embedding 해주기 위해 사용했습니다.
Graph Feature embedding은 아래의 세부 단계를 거치며 수행됩니다.

> 1. Macro Features와 Netlist Graph를 GNN에 입력하여 Node Embedding을 추출

Reviewer 2: 글 전체적으로  들여 쓰기(?) 된 문장에 대해 마침표를 붙일지 안 붙일지 통일이 되면 더 좋을 것 같습니다.
